---
title: "QTL Candidate Annotation for MR network"
output: html_notebook
---

## Background

Rob Baker et al fit growth parameters to B. rapa growth data and then used those to do function-value-trait QTL mapping and to find genes under those peaks

I used that same function-value-trait data to query a mutual rank (MR) gene expression network that I had built.  This returned genes that were in a networks associated with the growth paramters.  This was done separately for the CR and UN environments.  (Currently we are only focused on the UN environment)

Here I want to annotate candidate QTL genes with:

1. Are any network genes under the QTL peaks?  I have used Mutual Rank to define gene networks around the plant growth traits.  I am curious as to whether or not any of the genes in these networks underlie QTLs for these traits.

2. The degree of differential expression among the lines for the growth QTL genes.

```{r}
library(tidyverse)
library(magrittr)
library(limma)
```

## Annotation with MR network membership

```{r}
QTLgenes <- read_csv("../input/Bayesian2012Height_genesunderQTLS.csv")[,-1]
QTLgenes <- QTLgenes %>% filter(str_detect(FVTtrait,"^UN"))
QTLgenes
```

```{r}
MR_UN_genes <- read_csv("../output/MR_UN_graphs_node_annotation_2012.csv")
summary(MR_UN_genes)
MR_UN_genes
```

```{r}
QTLgenes <- MR_UN_genes %>%
  select(name, UN_network=.id, MR_Cutoff) %>%
  right_join(QTLgenes,by="name")
```

```{r}
QTLgenes %>% filter(!is.na(UN_network) ) %>% arrange(.id)
```


## Annotation with DE information

```{r}
load("../output/Voom_UN.Rdata")
```

```{r}
RIL.coefs.UN <- grep("RIL",colnames(coef(voom.fit.UN)))
RIL.pvals.UN <- topTable(voom.fit.UN,coef = RIL.coefs.UN,number = Inf)
```

```{r}
RIL.pvals.UN$rank_F <- rank(RIL.pvals.UN$F)

RIL.pvals.UN %<>% select(-starts_with("pdata"))
colnames(RIL.pvals.UN) %<>% paste(".UN",sep="")
```

```{r}
QTLgenes <- merge(QTLgenes, RIL.pvals.UN, by.x="name", by.y=0,all.x = TRUE, all.y = FALSE)
```

```{r}
write_csv(QTLgenes,"../output/Bayesian2012Height_genesunderQTLS_DE_MR_UN.csv")
```

Also subset to only include genes that are in an MR network

```{r}
QTLgenes  %>% filter(!is.na(UN_network) ) %>%
  write_csv("../output/Bayesian2012Height_genesunderQTLS_networkgenesonly_DE_MR_UN.csv")
```

## Test for enrichment

Determine number of genes within QTL boundaries, number of MR genes, and overlap, then Fisher's exact test.

Total genes
```{r}
annotation <- read_csv("../input/Brapa_V1.5_annotated.csv")
all.genes <- nrow(annotation)
all.genes
```

QTL genes
```{r}
qtl.genes <- nrow(QTLgenes)
qtl.genes
```

now go through each MR_cutoff

```{r}
fisher.results <- MR_UN_genes$MR_Cutoff %>%
  unique() %>%
  map(function(cutoff) {
    mr.genes <- sum(MR_UN_genes$MR_Cutoff <= cutoff)
    qtl.mr.genes <- sum(QTLgenes$MR_Cutoff <= cutoff,na.rm = TRUE)
    m <- matrix(c(all.genes-qtl.genes-mr.genes+qtl.mr.genes, qtl.genes-qtl.mr.genes,
             mr.genes-qtl.mr.genes,qtl.mr.genes),
           byrow=TRUE,
           ncol=2,
           dimnames=list(MR=c("no","yes"),QTL=c("no","yes"))
    )
    return(list(cutoff=cutoff,data=m,fisher=fisher.test(m)))
  }
  )

p.vals <- tibble(mr.cutoff=map_int(fisher.results,function(x) x[["cutoff"]]),
                fisher.p.val=map_dbl(fisher.results,function(x) x[["fisher"]]$p.value)
)

p.vals

lapply(fisher.results,function(x) x[["data"]])

combined.results <- t(sapply(fisher.results,function(x) x[["data"]])) %>% as.tibble()
colnames(combined.results) <- c("Genes_NoQTL_NoMR","Genes_NoQTL_YesMR","Genes_YesQTL_NoMR","Genes_YesQTL_YesMR")

combined.results <- cbind(p.vals,combined.results)

combined.results

write_csv(combined.results,
          path=str_c("../output/", "Bayesian2012Height_genesunderQTLS", "_MRoverlap_pvals_", Sys.Date(), ".csv") 
)
```
    
    
    
    
    