---
title: "Mutual Rank eQTL overlap"
output: html_notebook
---

## Background

Rob Baker et al fit growth parameters to B. rapa growth data and then used those to do function-value-trait QTL mapping.

I used those same parameters and queried a mutual rank (MR) gene expression network to find genes that were in a networks associated with these growth paramters.  This was done separeately for the CR and UN environments.  I looked for overlap between those MR-associated genes and growth parameters.

The goal now is to find the trans eQTL for each of the MR-growth associated genes and ask if the trans eQTL overlap with the growth/function-value QTL.  I think I am only going to take the top eQTL for each.

Focused on UN genes only.

## Methodology

For each gene in a MR network with a growth parameter query the eQTL database to find its trans eQTL regions.  Then compare overlaps between those and the growth QTL.

## Load the libraries and data

Libraries
```{r}
library(tidyverse)
library(stringr)
library(magrittr)
library(qtl)
```

Growth QTL
```{r}
QTLgenes <- read_csv("../input/BrapaNetworks/Heigh2012_uniquegenes.csv")[,-1]
summary(QTLgenes)
QTLgenes
```

MR genes from UN
```{r}
MR_UN_genes <- read_csv("../output/MR_UN_graphs_annotation.csv") %>%
  filter(MR_Cutoff <= 50, !duplicated(name)) %>%
  select(MR_Cutoff,name)
MR_UN_genes
```

eQTL
```{r}
load("../brassica_eqtl_paper/input/scanone-eqtl-UN.Rdata")
scanone_imp_UN <- scanone_imp_un
```

```{r}
ls()
```

## Find eQTL intervals


### UN data

Get the scanone data in a nice format for summarizing and plotting
```{r}
scanone_UN_gather <- scanone_imp_UN %>%
  gather(key = gene, value = LOD, -chr, -pos) %>%
  semi_join(MR_UN_genes,by=c("gene"="name")) # only keep genes in MR networks
```

plot eQTL peaks...
```{r, fig.height=10}
pl.UN <- scanone_UN_gather %>%
  ggplot(aes(x=pos,y=LOD,color=gene)) +
  geom_line() +
  geom_hline(aes(yintercept=4),lty=2,lwd=.5,alpha=.5) +
  facet_grid( ~ chr, scales="free") +
  theme(strip.text.y = element_text(angle=0), axis.text.x = element_text(angle=90)) +
  ggtitle("MR gene eQTL")
pl.UN
ggsave("MR gene eQTL UN.pdf",width=12,height=8)
```


```{r}
sig_chromosomes_UN <- scanone_UN_gather %>%
  group_by(gene,chr) %>%
  summarize(pos=pos[which.max(LOD)],LOD=max(LOD)) %>%
  filter(LOD > 4)

sig_chromosomes_UN
```

now for each significant chromosome/trait combo run bayesint

```{r}
bayesint_list_UN <- apply(sig_chromosomes_UN,1,function(hit) {
  result <- bayesint(scanone_imp_UN[c("chr","pos",hit["gene"])], 
                     chr=hit["chr"], 
                     lodcolumn = 1,
                     expandtomarkers = TRUE
  )
  colnames(result)[3] <- "LOD"
  result
})
names(bayesint_list_UN) <- sig_chromosomes_UN$gene

bayesint_list_UN <- lapply(bayesint_list_UN,function(x) x %>% 
                             as.data.frame(stringsAsFactors=FALSE) %>%
                             rownames_to_column(var="markername") %>%
                             mutate(chr=as.character(chr))
)
```

```{r}
bayesint_result_UN <- as.tibble(bind_rows(bayesint_list_UN,.id="gene")) %>% 
  select(gene,chr,pos,markername,LOD) %>%
  separate(markername,into=c("chr1","Mbp"),sep="x", convert=TRUE) %>%
  group_by(gene,chr) %>%
  summarize(start=min(Mbp),end=max(Mbp),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD)) %>%
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))

bayesint_result_UN
```

## annotate UN eQTL

Load annotation
```{r}
BrapaAnnotation <- read_csv("../input/Brapa_V1.5_annotated.csv")
BrapaAnnotation
```


```{r}
UN_annotated <- lapply(1:nrow(bayesint_result_UN),function(row) {
  qtl <- bayesint_result_UN[row,]
  subset(BrapaAnnotation, chrom==qtl$chr &
           start >= qtl$start &
           end <= qtl$end) 
}
)
names(UN_annotated) <- bayesint_result_UN$gene
UN_annotated <- bind_rows(UN_annotated,.id="MR_gene") %>%
  left_join(bayesint_result_UN,by=c("MR_gene"="gene","chrom"="chr")) %>% #get eQTL LOD
    left_join(MR_UN_genes,by=c("MR_gene"="name")) %>% # get cutoff
  rename(eQTL_candidate=name)


UN_annotated_small <- UN_annotated %>% select(MR_gene,MR_Cutoff,eQTL_candidate,ends_with("LOD"))

UN_annotated_small
```

given bayesint results, find overlaps with UN growth QTL

```{r}
UN_MReQTL_QTL_combined <- inner_join(QTLgenes,UN_annotated_small,by=c("name"="eQTL_candidate")) %>%
  select(.id, MR_gene, MR_Cutoff, ends_with("LOD"), everything()) %>%
  arrange(.id,desc(max_eQTL_LOD)) %>%
  rename(eQTL_candidate=name)
UN_MReQTL_QTL_combined

UN_MReQTL_QTL_combined_small <- UN_MReQTL_QTL_combined %>% filter(!duplicated(eQTL_candidate)) %>%
  select(-MR_gene,-MR_Cutoff) 
UN_MReQTL_QTL_combined_small
```

how many QTL have at least some overlap?
```{r}
unique(QTLgenes$.id)
unique(UN_MReQTL_QTL_combined$.id)
```

ALL

```{r}
write_csv(UN_MReQTL_QTL_combined,"../output/Heigh2012_uniquegenes_MR_eQTL_UN_overlap.csv")
```





