---
title: "Analyze Eigen Gene QTL"
output: html_notebook
author: "Julin Maloof"
---


Previous script: "06_get_eigengene_QTL.Rmd"

The goal is to find QTL peaks for the WGCNA eigen genes and see if those overalp with any growth QTL.  We are only focusing on eigen genes that correlated with some growth traits/paramters.

```{r}
library(GenomicRanges)
library(qtl)
library(tidyverse)
library(stringr)
load("../output/scanone-eigengene-qtl_2012.RData")

```

# cim

## Plot QTL

```{r}

threshold.95 <- permtest.cim %>% apply(1,max) %>% quantile(.95)
threshold.95

scanone.gather <- scanone_eigen_cim %>%
  gather(key = trait, value = LOD, -chr, -pos) %>%
  mutate(condition=str_sub(trait,1,2), 
         color=str_sub(trait,6,100),
         perm.threshold=threshold.95)

scanone.gather
```


```{r}
   pl.UN <- scanone.gather %>% filter(condition=="UN") %>%
  ggplot(aes(x=pos,y=LOD)) +
  geom_line() +
  geom_hline(aes(yintercept=perm.threshold),lty=2,lwd=.5,alpha=.5) +
  facet_grid(trait ~ chr, scales="free") +
  theme(strip.text.y = element_text(angle=0), axis.text.x = element_text(angle=90)) +
  ggtitle("UN Eigen Gene QTL")
pl.UN
ggsave("../output/eigen gene eQTL UN CIM 2012.pdf",width=12,height=8)
```


## Look for overlap

For each eigen gene, find QTL borders and look for overlap with growth QTL

For each eigen gene first identify chromosomes with "significant" peaks (in this case > 95% permuation threshold) and then run bayesint() on them to define the intervals

```{r}
sig.chrs <- scanone.gather %>% filter(LOD > perm.threshold) %>%
  group_by(trait,chr) %>%
  summarize(unique(chr))
sig.chrs
```

now for each significant chromosome/trait combo run bayesint

```{r}
#remove markers without physical position
scanone_eigen_cim.phys <- scanone_eigen_cim[!str_detect(rownames(scanone_eigen),"^cA"),]

bayesint.list <- apply(sig.chrs,1,function(hit) {
    result <- bayesint(scanone_eigen_cim.phys[c("chr","pos",hit["trait"])], 
                     chr=hit["chr"], 
                     lodcolumn = 1,
                     expandtomarkers = TRUE
  )
  colnames(result)[3] <- "LOD"
  result
})

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x) x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>%
                          mutate(chr=as.character(chr))
)

bayesint.result <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% 
  select(trait,chr,pos,markername,LOD) %>%
  separate(markername,into=c("chr1","Mbp"),sep="x", convert=TRUE) %>%
  group_by(trait,chr) %>% 
  summarize(eQTL_start_bp=min(Mbp), eQTL_end_bp=max(Mbp), eQTL_start_cM=min(pos), eQTL_end_cM=max(pos), min_eQTL_LOD=min(LOD), max_eQTL_LOD=max(LOD)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(expand_interval=eQTL_start_cM==eQTL_end_cM,
         eQTL_start_cM=ifelse(expand_interval ,max(0, eQTL_start_cM - 2.5), eQTL_start_cM), 
         eQTL_end_cM=ifelse(expand_interval, eQTL_end_cM + 2.5, eQTL_end_cM),
         eQTL_start_bp=ifelse(expand_interval,
                              {find.flanking(cross.all, chr, eQTL_start_cM) %>% 
                                  pull(left) %>% 
                                  str_remove("^A[01][0-9]x") %>% 
                                  as.numeric()},
                              eQTL_start_bp),
         eQTL_end_bp=ifelse(expand_interval,
                              {find.flanking(cross.all, chr, eQTL_end_cM) %>% 
                                  pull(left) %>% 
                                  str_remove("^A[01][0-9]x") %>% 
                                  as.numeric()},
                              eQTL_end_bp))  
  
bayesint.result %>% arrange(trait)
bayesint.result %>% arrange(chr)
```

### annotate Eigen gene QTL

Load annotation
```{r}
BrapaAnnotation <- read_csv("../input/Brapa_V1.5_annotated.csv")
BrapaAnnotation
```

```{r}
eigen.annotated <- lapply(1:nrow(bayesint.result),function(row) {
  qtl <- bayesint.result[row,]
  subset(BrapaAnnotation, chrom==qtl$chr &
           start >= qtl$eQTL_start_bp &
           end <= qtl$eQTL_end_bp) 
}
)
names(eigen.annotated) <- bayesint.result$trait

eigen.annotated <- bind_rows(eigen.annotated,.id="trait") %>%
  mutate(chrom=as.character(chrom)) %>%
  left_join(bayesint.result,by=c("trait","chrom"="chr")) %>% #get eQTL LOD
  rename(eigen_eQTL_candidate=name)

eigen.annotated.small <- eigen.annotated %>% select(trait,eigen_eQTL_candidate,ends_with("LOD"))

eigen.annotated.small
```

given bayesint results, find overlaps with UN growth QTL

```{r}
filepath <- "../input/All2012HeightQTL2.xlsx"
filebase <- filepath %>% basename() %>% str_replace("\\..*$","")

QTLgenes <- readxl::read_excel(filepath)[,-1]
QTLgenes <- QTLgenes %>% dplyr::rename(.id=QTL, FVTtrait=FVT) # change names to match previous file
QTLgenes <- QTLgenes %>% filter(str_detect(FVTtrait,"^UN"))
QTLgenes
```

```{r}
eigen.qtl.combined <- inner_join(eigen.annotated.small,QTLgenes,by=c("eigen_eQTL_candidate"="name")) %>%
  select(.id, trait, everything())
eigen.qtl.combined
```

how many QTL have at least some overlap?
```{r}
unique(QTLgenes$.id)
unique(eigen.qtl.combined$.id)
```

five of 16

are all eigen genes overlapping?

```{r}
unique(eigen.annotated.small$trait)
unique(eigen.qtl.combined$trait)
```

No, 4 of 6

```{r}
write_csv(eigen.qtl.combined,
          path=str_c("../output/", filebase, "_eigenQTL_overlap_CIM_", Sys.Date(), ".csv"))
```

## overlaps and signficance

Make table of chromosome info
```{r}
chr.info.cim <- scanone_eigen_cim.phys %>% 
  as.data.frame() %>%
  rownames_to_column("marker") %>%
  select(marker) %>%
  separate(marker,into=c("chr","bp"),sep="x",convert=TRUE) %>%
  group_by(chr) %>%
  summarize(start=min(bp),end=max(bp))
```

```{r}
qtl.info <- QTLgenes %>%
  group_by(.id) %>%
  summarize(chrom=unique(chrom),start=min(start),end=max(end))
qtl.info
qtl.ranges <- GRanges(seqnames = qtl.info$chrom,ranges=IRanges(start=qtl.info$start,end=qtl.info$end))
qtl.ranges <- GenomicRanges::reduce(qtl.ranges)

eQTL.ranges <- GRanges(bayesint.result$chr,
                       ranges = IRanges(start=bayesint.result$eQTL_start_bp,
                                        end=bayesint.result$eQTL_end_bp))
eQTL.ranges <- GenomicRanges::reduce(eQTL.ranges)
```

Do the simulations
```{r}
sims <- 1000

set.seed(4545)
sim.results <- sapply(1:sims, function(s) {
  if (s %% 100 == 0) print(s)
  sim.eQTL <- tibble(
    chr=sample(chr.info.cim$chr,
               size = length(eQTL.ranges),
               replace = TRUE,
               prob=chr.info.cim$end/sum(chr.info.cim$end)),
    width=width(eQTL.ranges) # width of the QTL to simulate
  )
  sim.eQTL <- chr.info.cim %>% 
    select(chr,chr.start=start,chr.end=end) %>% right_join(sim.eQTL,by="chr") #need to get the chrom end so we can sample correctly
  sim.eQTL <- sim.eQTL %>% mutate(qtl.start = runif(n=n(),
                                                    min = chr.start,
                                                    max= max(chr.start,chr.end-width)),
                                  qtl.end=qtl.start+width)
  sim.eQTL.ranges <- GRanges(seqnames = sim.eQTL$chr,
                             ranges = IRanges(start=sim.eQTL$qtl.start,
                                              end=sim.eQTL$qtl.end))
  
  suppressWarnings(result <- sum(countOverlaps(qtl.ranges,sim.eQTL.ranges)>0))
  result
})

```


```{r}
true.overlap <- sum(countOverlaps(qtl.ranges,eQTL.ranges)) #OK to ignore warnings

true.overlap

sum(sim.results >= true.overlap)

mean(sim.results >= true.overlap)

tibble(FVTQTL_vs_MReQTL_True_Overlaps=true.overlap,
       N_Simulations_fewer_overlaps=sum(sim.results < true.overlap),
       N_Simulations_greater_equal_overlaps=sum(sim.results >= true.overlap),
       P_value=mean(sim.results >= true.overlap)
) %>%
  write_csv(str_c("../output/", filebase, "_WGCNA_eigen_eQTL_CIM_overlap_pval_", Sys.Date(), ".csv"))
```


